codespell_lib/tests/test_dictionary.py

```diff
@@ -1,61 +1,152 @@
 # -*- coding: utf-8 -*-
 
+import glob
 import os.path as op
+import os
 import re
 
+import pytest
 
-def test_dictionary_formatting():
+from codespell_lib._codespell import _builtin_dictionaries
+
+try:
+    import aspell
+    speller = aspell.Speller('lang', 'en')
+except Exception as exp:  # probably ImportError, but maybe also language
+    speller = None
+    if os.getenv('ALLOW_NO_ASPELL', 'false').lower() != 'true':
+        raise RuntimeError(
+            'Cannot run complete tests without aspell, consider installing '
+            'it, or set ALLOW_NO_ASPELL=true. Got error during import:\n%s'
+            % (exp,))
+
+ws = re.compile(r'.*\s.*')  # whitespace
+comma = re.compile(r'.*,.*')  # comma
+
+
+# Filename, should be seen as errors in aspell or not
+_data_dir = op.join(op.dirname(__file__), '..', 'data')
+_fnames_in_aspell = [
+    (op.join(_data_dir, 'dictionary%s.txt' % d[2]), d[3:5])
+    for d in _builtin_dictionaries]
+fname_params = pytest.mark.parametrize('fname, in_aspell', _fnames_in_aspell)
+
+
+def test_dictionaries_exist():
+    """Test consistency of dictionaries."""
+    doc_fnames = set(op.basename(f[0]) for f in _fnames_in_aspell)
+    got_fnames = set(op.basename(f)
+                     for f in glob.glob(op.join(_data_dir, '*.txt')))
+    assert doc_fnames == got_fnames
+
+
+@fname_params
+def test_dictionary_formatting(fname, in_aspell):
     """Test that all dictionary entries are valid."""
-    err_dict = dict()
-    ws = re.compile(r'.*\s.*')  # whitespace
-    comma = re.compile(r'.*,.*')  # comma
-    with open(op.join(op.dirname(__file__), '..', 'data',
-                      'dictionary.txt'), 'rb') as fid:
+    errors = list()
+    with open(fname, 'rb') as fid:
         for line in fid:
             err, rep = line.decode('utf-8').split('->')
             err = err.lower()
             rep = rep.rstrip('\n')
-            assert err != rep.lower(), 'error %r corrects to itself' % err
+            try:
+                _check_err_rep(err, rep, in_aspell, fname)
+            except AssertionError as exp:
+                errors.append(str(exp).split('\n')[0])
+    if len(errors):
+        raise AssertionError('\n' + '\n'.join(errors))
+
+
+def _check_aspell(word, msg, in_aspell, fname):
+    if speller is None:
+        return  # cannot check
+    if in_aspell is None:
+        return  # don't check
+    if ' ' in word:
+        return  # can't check (easily)
+    this_in_aspell = speller.check(
+        word.encode(speller.ConfigKeys()['encoding'][1]))
+    end = 'be in aspell for dictionary %s' % (fname,)
+    if in_aspell:  # should be an error in aspell
+        assert this_in_aspell, '%s should %s' % (msg, end)
+    else:  # shouldn't be
+        assert this_in_aspell, '%s should not %s' % (msg, end)
```

Comment:
Yes it's possible, but I find it less clear actually

Reply:
Fair enough.

Reply:
I thought this smelt a bit funny @larsoner and having added some tests, I realised why, it's because the logic for the not in aspell case wasn't right! Now fixed and in the test, I'm just working on updating the actual dictionary tests.